File: ruby/scrabble-score/scrabble_score.rb
# give score for a word on Scrabble game
class Scrabble
  TILES = {
    A: 1, N: 1,
    B: 3, O: 1,
    C: 3, P: 3,
    D: 2, Q: 10,
    E: 1, R: 1,
    F: 4, S: 1,
    G: 2, T: 1,
    H: 4, U: 1,
    I: 1, V: 4,
    J: 8, W: 4,
    K: 5, X: 8,
    L: 1, Y: 4,
    M: 3, Z: 10
  }

  def self.score(tiles)
    new(tiles).score
  end

  private

  attr_reader :letters, :tiles

  def initialize(word, tiles = TILES)
    @letters = word.to_s.upcase.chars.map(&:to_sym)
    tiles.default = 0
    @tiles = tiles
  end

  public

  def score
    letters.sum(&tiles)
  end
end

File: erlang/scrabble-score/src/scrabble_score.erl
-module(scrabble_score).

-export([score/1]).

do_score([$A | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$E | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$I | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$O | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$U | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$L | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$N | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$R | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$S | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$T | T], Acc) ->
    do_score(T, Acc + 1);
do_score([$D | T], Acc) ->
    do_score(T, Acc + 2);
do_score([$G | T], Acc) ->
    do_score(T, Acc + 2);
do_score([$B | T], Acc) ->
    do_score(T, Acc + 3);
do_score([$C | T], Acc) ->
    do_score(T, Acc + 3);
do_score([$M | T], Acc) ->
    do_score(T, Acc + 3);
do_score([$P | T], Acc) ->
    do_score(T, Acc + 3);
do_score([$F | T], Acc) ->
    do_score(T, Acc + 4);
do_score([$H | T], Acc) ->
    do_score(T, Acc + 4);
do_score([$V | T], Acc) ->
    do_score(T, Acc + 4);
do_score([$W | T], Acc) ->
    do_score(T, Acc + 4);
do_score([$Y | T], Acc) ->
    do_score(T, Acc + 4);
do_score([$K | T], Acc) ->
    do_score(T, Acc + 5);
do_score([$J | T], Acc) ->
    do_score(T, Acc + 8);
do_score([$X | T], Acc) ->
    do_score(T, Acc + 8);
do_score([$Q | T], Acc) ->
    do_score(T, Acc + 10);
do_score([$Z | T], Acc) ->
    do_score(T, Acc + 10);
do_score([], Acc) ->
    Acc.

score(Word) ->
    do_score(string:uppercase(Word), 0).

